File: PaymentezSDKObjC/PaymentezSDKObjC/ListCardsTableViewController.m
//
//  ListCardsTableViewController.m
//  PaymentezSDKObjC
//
//  Created by Gustavo Sotelo on 10/05/16.
//  Copyright © 2016 Paymentez. All rights reserved.
//

#import "ListCardsTableViewController.h"
#import "ViewController.h"
#import <PaymentezSDK/PaymentezSDK-Swift.h>

@interface ListCardsTableViewController ()
@property (nonatomic, strong) NSMutableArray *cardList;
@end

@implementation ListCardsTableViewController


- (NSMutableArray*) getCardList
{
    if (_cardList == nil)
    {
        _cardList = [[NSMutableArray alloc] init];
    }
    return _cardList;
}
- (IBAction)addAction:(id)sender {
    
    //Show an activity indicator
    [PaymentezSDKClient showAddViewControllerForUser:@"gus" email:@"gsotelo@paymentez.com" presenter:self callback:^(PaymentezSDKError *error, BOOL closed, BOOL added) {
        if (error != nil)
            NSLog(@"%@", error.description);
        else if (closed)
            NSLog(@"User closed the modal");
        else if (added)
            NSLog(@"Card Added ");
            
    }];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self refreshTable];
}

- (void) refreshTable
{
    [PaymentezSDKClient listCards:@"gus" callback:^(PaymentezSDKError * error, NSArray<PaymentezCard *> * list) {
        if(error == nil)
        {
            self.cardList = [list mutableCopy];
            
            [self.tableView reloadData];
        }
    } ];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
#warning Incomplete implementation, return the number of sections
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
#warning Incomplete implementation, return the number of rows
    return self.cardList.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cardCell" forIndexPath:indexPath];
    PaymentezCard *card = self.cardList[indexPath.row];
    cell.textLabel.text = card.termination;
    cell.detailTextLabel.text = card.cardReference;
    
    // Configure the cell...
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    PaymentezCard *card = self.cardList[indexPath.row];
    self.cardReference = card.cardReference;
    [self performSegueWithIdentifier:@"debitSegue" sender:self];
    
}
/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle: (UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        PaymentezCard *card = self.cardList[indexPath.row];
        
        [PaymentezSDKClient deleteCard:@"gus" cardReference:card.cardReference callback:^(PaymentezSDKError *error, BOOL deleted) {
           
            [self refreshTable];
            
            
        }];
        
    }
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    
    if ([segue.identifier isEqualToString:@"debitSegue"])
    {
    ViewController *vc = (ViewController*)segue.destinationViewController;
    vc.cardReference  = self.cardReference;
    }
}


@end

File: PaymentezSDKSwift/PaymentezSDKSwift/ListCardsTableViewController.swift
//
//  ListCardsTableViewController.swift
//  PaymentezSDKSample
//
//  Created by Gustavo Sotelo on 09/05/16.
//  Copyright © 2016 Paymentez. All rights reserved.
//

import UIKit
import PaymentezSDK

protocol CardSelectedDelegate
{
    func cardSelected(card:PaymentezCard?)
}

class ListCardsTableViewController: UITableViewController {

    var cardList = [PaymentezCard]()
    var cardSelectedDelegate:CardSelectedDelegate?
    
    override func viewDidLoad() {
        super.viewDidLoad()
                // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.refreshTable()

    }
    func refreshTable()
    {
        self.cardList.removeAll()
        
        MyBackendLib.listCard(uid: "1234") { (cardList) in
            self.cardList = cardList!
            self.tableView.reloadData();
        }
        
        
        
        /*PaymentezSDKClient.listCards(UserModel.uid) { (error, cardList) in
            
            if error == nil
            {
                self.cardList = cardList!
                
                
            }
            
            
        }*/
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        // #warning Incomplete implementation, return the number of sections
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return cardList.count
    }

    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cardCell", for: indexPath)
        
        let card = self.cardList[(indexPath as NSIndexPath).row]
        cell.textLabel!.text = "****" + card.termination!
        cell.detailTextLabel!.text = "\(card.expiryMonth!)/\(card.expiryYear!)"
        cell.imageView?.image = card.getCardTypeAsset()
        // Configure the cell...

        return cell
    }
 
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let card  = self.cardList[(indexPath as NSIndexPath).row]
        self.cardSelectedDelegate?.cardSelected(card: card)
        self.navigationController?.popViewController(animated: true)
        //self.performSegue(withIdentifier: "debitSegue", sender: self)
        
    }
    
    // MARK: - Navigation

    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    override func tableView(_ tableView: UITableView, editActionsForRowAt indexPath: IndexPath) -> [UITableViewRowAction]? {
        
        let card = self.cardList[(indexPath as NSIndexPath).row]
        var actions = [UITableViewRowAction]()
        if card.status == "review"
        {
            let editAction = UITableViewRowAction(style: .normal, title: "Verify amount") { (rowAction, indexPath) in
                
                //DEV BACKEND WITH VERIFY WITH AMOUNT
            }
            editAction.backgroundColor = .blue
            
            let editAction2 = UITableViewRowAction(style: .normal, title: "Verify code") { (rowAction, indexPath) in
               
                //DEV BACKEND WITH VERIFY WITH CODE
                
            }
            editAction2.backgroundColor = .gray
            actions.append(editAction)
            actions.append(editAction2)
        }
        let deleteAction = UITableViewRowAction(style: .normal, title: "Delete") { (rowAction, indexPath) in
            let card = self.cardList[(indexPath as NSIndexPath).row]
            
            //DEV BACKEND CALL WITH DELETE CARD

        }
        deleteAction.backgroundColor = .red
        actions.append(deleteAction)
        return actions
    }
    
   
    
    
}

